mlir/lib/Analysis/BufferDependencyAnalysis.cpp

```diff
@@ -0,0 +1,153 @@
+  for (auto *user : parent->getUsers()) {
+    if (auto viewOp = dyn_cast<ViewLikeOpInterface>(user)) {
+      findAllUsers(viewOp, users);
+    } else {
+      users.push_back({parent, user});
+    }
+  }
+}
+
+enum EffectType { read, write, unknown };
+EffectType getMemoryEffect(UserWrapper &user) {
+  if (auto memoryEffect = dyn_cast<MemoryEffectOpInterface>(user.op)) {
+    int64_t index = -1;
+    for (auto it : llvm::enumerate(user.op->getOperands())) {
+      auto operand = it.value();
+      if (operand.getDefiningOp() == user.parentAlias) {
+        index = static_cast<int64_t>(it.index());
+        break;
+      }
+    }
+    if (index < 0)
+      return EffectType::unknown;
+    Value value = user.op->getOperand(index);
+    if (memoryEffect.getEffectOnValue<MemoryEffects::Write>(value)) {
+      return EffectType::write;
+    }
+    if (memoryEffect.getEffectOnValue<MemoryEffects::Read>(value)) {
+      return EffectType::read;
+    }
+  }
+  return EffectType::unknown;
+}
+
+template <typename Impl>
+void assign(EffectType effect, UserWrapper &user, Impl &impl) {
+  if (effect == Impl::type) {
+    impl.push_back(user.op);
+  }
+}
+
+template <typename Impl, typename... Impls>
+void assign(EffectType effect, UserWrapper &user, Impl &impl, Impls &...impls) {
+  assign(effect, user, impl);
+  assign(effect, user, impls...);
+}
+
+template <typename Impl, typename... Impls>
+void find(memref::AllocOp allocOp, Impl &impl, Impls &...impls) {
+  BufferDependencyAnalysis::Pair result;
+
+  llvm::SmallVector<UserWrapper> users;
+  findAllUsers(allocOp, users);
+  for (auto &user : users) {
+    EffectType memoryEffect = getMemoryEffect(user);
+    assign(memoryEffect, user, impl, impls...);
+  }
+}
+
+struct Readers : public llvm::SmallVector<Operation *> {
+  static inline constexpr EffectType type = EffectType::read;
+};
+
+struct Writers : public llvm::SmallVector<Operation *> {
+  static inline constexpr EffectType type = EffectType::write;
+};
+
+BufferDependencyAnalysis::BufferDependencyAnalysis(Operation *op) {
+  func = dyn_cast<func::FuncOp>(op);
+  assert(func &&
+         "expected func::FuncOp as the top level operation for the analysis");
+}
+
+llvm::SmallVector<Operation *>
+BufferDependencyAnalysis::getReaders(memref::AllocOp allocOp) {
+  Readers readers;
+  find(allocOp, readers);
+  return std::move(readers);
+}
+
+llvm::SmallVector<Operation *>
+BufferDependencyAnalysis::getWriters(memref::AllocOp allocOp) {
+  Writers writers;
+  find(allocOp, writers);
+  return std::move(writers);
+}
+
+BufferDependencyAnalysis::Pair
+BufferDependencyAnalysis::getReadersAndWriters(memref::AllocOp allocOp) {
+  Readers readers;
+  Writers writers;
+  find(allocOp, readers, writers);
+
+  BufferDependencyAnalysis::Pair result;
+  result.readers = std::move(readers);
+  result.writers = std::move(writers);
+
+  return result;
+}
+
+std::optional<memref::AllocOp>
+BufferDependencyAnalysis::getAllocation(Value value) {
```

Comment:
This should be cached in the analysis.

Reply:
... and part of me is tempted to have a nice table that also gives you all the views, so that `untransform()` is a analysis lookup, but that's a future PR.